File: src/common/events.ts
export enum SchaltEvent {
  SessionsRefreshed = 'schaltwerk:sessions-refreshed',
  SessionAdded = 'schaltwerk:session-added',
  SessionRemoved = 'schaltwerk:session-removed',
  ArchiveUpdated = 'schaltwerk:archive-updated',
  SessionCancelling = 'schaltwerk:session-cancelling',
  CancelError = 'schaltwerk:cancel-error',
  TerminalCreated = 'schaltwerk:terminal-created',

  SessionActivity = 'schaltwerk:session-activity',
  SessionGitStats = 'schaltwerk:session-git-stats',
  TerminalAttention = 'schaltwerk:terminal-attention',
  TerminalClosed = 'schaltwerk:terminal-closed',
  TerminalAgentStarted = 'schaltwerk:terminal-agent-started',
  TerminalForceScroll = 'schaltwerk:terminal-force-scroll',
  GlobalKeepAwakeStateChanged = 'schaltwerk:global-keep-awake-state-changed',
  PtyData = 'schaltwerk:pty-data',
  ProjectReady = 'schaltwerk:project-ready',
  OpenDirectory = 'schaltwerk:open-directory',
  OpenHome = 'schaltwerk:open-home',
  FileChanges = 'schaltwerk:file-changes',
  FollowUpMessage = 'schaltwerk:follow-up-message',
  Selection = 'schaltwerk:selection',
  GitOperationStarted = 'schaltwerk:git-operation-started',
  GitOperationCompleted = 'schaltwerk:git-operation-completed',
  GitOperationFailed = 'schaltwerk:git-operation-failed',
  ProjectFilesUpdated = 'schaltwerk:project-files-updated',
  GitHubStatusChanged = 'schaltwerk:github-status-changed',
  GitLabStatusChanged = 'schaltwerk:gitlab-status-changed',
  AppUpdateResult = 'schaltwerk:app-update-result',
  DevBackendError = 'schaltwerk:dev-backend-error',
  SetupScriptRequested = 'schaltwerk:setup-script-request',
  CloneProgress = 'schaltwerk:clone-progress',
  OrchestratorLaunchFailed = 'schaltwerk:orchestrator-launch-failed',
  DiffBaseBranchChanged = 'schaltwerk:diff-base-branch-changed',
  ProjectValidationError = 'schaltwerk:project-validation-error',
  OpenPrModal = 'schaltwerk:open-pr-modal',
  OpenMergeModal = 'schaltwerk:open-merge-modal',
  OpenGitlabMrModal = 'schaltwerk:open-gitlab-mr-modal',
  SelectAllRequested = 'schaltwerk:select-all-requested',
}


export interface SessionActivityUpdated {
  session_id: string
  session_name: string
  last_activity_ts: number
  current_task: string | null
  todo_percentage: number | null
  is_blocked: boolean | null
}

export interface SessionGitStatsUpdated {
  session_id: string
  session_name: string
  files_changed: number
  lines_added: number
  lines_removed: number
  has_uncommitted: boolean
  has_conflicts?: boolean
  top_uncommitted_paths?: string[]
  merge_has_conflicts?: boolean
  merge_conflicting_paths?: string[]
  merge_is_up_to_date?: boolean
}

export interface FollowUpMessagePayload {
  session_name: string
  message: string
  timestamp: number
  terminal_id: string
  message_type: 'system' | 'user'
}

export interface PtyDataPayload {
  term_id: string
  seq: number
  base64: string
}

export interface GlobalKeepAwakeStatePayload {
  state: 'disabled' | 'active' | 'auto_paused'
  activeCount?: number
}

export interface ChangedFile {
  path: string
  change_type: 'modified' | 'added' | 'deleted' | 'renamed' | 'copied' | 'unknown'
  additions: number
  deletions: number
  changes: number
  is_binary?: boolean
  previous_path?: string
}

export interface BranchInfo {
  current_branch: string
  base_branch: string
  base_commit: string
  head_commit: string
}

export interface GitOperationPayload {
  session_name: string
  session_branch: string
  parent_branch: string
  mode: string
  operation: 'merge'
  commit?: string
  status?: 'started' | 'success' | 'conflict' | 'error'
}

export interface GitOperationFailedPayload extends GitOperationPayload {
  error: string
}

export interface GitHubRepositoryPayload {
  nameWithOwner: string
  defaultBranch: string
}

export interface GitHubStatusPayload {
  installed: boolean
  authenticated: boolean
  userLogin?: string | null
  repository?: GitHubRepositoryPayload | null
}

export interface GitLabStatusPayload {
  installed: boolean
  authenticated: boolean
  userLogin?: string | null
  hostname?: string | null
}

export interface GitHubPrPayload {
  branch: string
  url: string
}

export type UpdateResultStatus = 'updated' | 'upToDate' | 'error' | 'busy'
export type UpdateInitiator = 'auto' | 'manual'
export type UpdateErrorKind = 'network' | 'permission' | 'signature' | 'unknown'

export interface AppUpdateResultPayload {
  status: UpdateResultStatus
  initiatedBy: UpdateInitiator
  currentVersion: string
  newVersion?: string
  notes?: string | null
  errorKind?: UpdateErrorKind
  errorMessage?: string
}

export interface DevBackendErrorPayload {
  message: string
  source?: string
}

export type CloneProgressKind = 'info' | 'success' | 'error'

export interface CloneProgressPayload {
  requestId: string
  message: string
  remote: string
  kind: CloneProgressKind
}

export interface SetupScriptRequestPayload {
  setup_script: string
  has_setup_script: boolean
  project_path?: string
  pending_confirmation?: boolean
}

export interface OrchestratorLaunchFailedPayload {
  terminal_id: string
  error: string
}

export interface DiffBaseBranchChangedPayload {
  session_name: string
  new_base_branch: string
}

export interface ProjectValidationErrorPayload {
  path: string
  error: string
}

export interface OpenPrModalPayload {
  sessionName: string
  prTitle?: string
  prBody?: string
  baseBranch?: string
  prBranchName?: string
  mode?: 'squash' | 'reapply'
}

export interface OpenMergeModalPayload {
  sessionName: string
  mode?: 'squash' | 'reapply'
  commitMessage?: string
}

export interface OpenGitlabMrModalPayload {
  sessionName: string
  suggestedTitle?: string
  suggestedBody?: string
  suggestedBaseBranch?: string
  suggestedSourceProject?: string
}

import { type EnrichedSession, type Epic } from '../types/session'

export interface SessionsRefreshedEventPayload {
  projectPath: string
  sessions: EnrichedSession[]
}

export interface SelectionPayload {
  kind: 'session' | 'orchestrator'
  payload?: string
  worktreePath?: string
  sessionState?: 'spec' | 'processing' | 'running' | 'reviewed'
}

export type EventPayloadMap = {
  [SchaltEvent.SessionsRefreshed]: SessionsRefreshedEventPayload
  [SchaltEvent.SessionAdded]: {
    session_name: string
    branch: string
    worktree_path: string
    parent_branch: string
    created_at: string
    last_modified?: string
    epic?: Epic
    agent_type?: string
    skip_permissions?: boolean
  }
  [SchaltEvent.SessionRemoved]: { session_name: string }
  [SchaltEvent.ArchiveUpdated]: { repo: string, count: number }
  [SchaltEvent.SessionCancelling]: { session_name: string }
  [SchaltEvent.CancelError]: { session_name: string, error: string }
  [SchaltEvent.TerminalCreated]: { terminal_id: string, cwd: string }

  [SchaltEvent.SessionActivity]: SessionActivityUpdated
  [SchaltEvent.SessionGitStats]: SessionGitStatsUpdated
  [SchaltEvent.TerminalAttention]: { session_id: string, terminal_id: string, needs_attention: boolean }
  [SchaltEvent.TerminalClosed]: { terminal_id: string }
  [SchaltEvent.TerminalAgentStarted]: { terminal_id: string, session_name?: string }
  [SchaltEvent.TerminalForceScroll]: { terminal_id: string }
  [SchaltEvent.GlobalKeepAwakeStateChanged]: GlobalKeepAwakeStatePayload
  [SchaltEvent.PtyData]: PtyDataPayload
  [SchaltEvent.ProjectReady]: string
  [SchaltEvent.OpenDirectory]: string
  [SchaltEvent.OpenHome]: string
  [SchaltEvent.FileChanges]: {
    session_name: string
    changed_files: ChangedFile[]
    branch_info: BranchInfo
  }
  [SchaltEvent.FollowUpMessage]: FollowUpMessagePayload
  [SchaltEvent.Selection]: SelectionPayload
  [SchaltEvent.GitOperationStarted]: GitOperationPayload
  [SchaltEvent.GitOperationCompleted]: GitOperationPayload
  [SchaltEvent.GitOperationFailed]: GitOperationFailedPayload
  [SchaltEvent.ProjectFilesUpdated]: string[]
  [SchaltEvent.GitHubStatusChanged]: GitHubStatusPayload
  [SchaltEvent.GitLabStatusChanged]: GitLabStatusPayload
  [SchaltEvent.AppUpdateResult]: AppUpdateResultPayload
  [SchaltEvent.DevBackendError]: DevBackendErrorPayload
  [SchaltEvent.SetupScriptRequested]: SetupScriptRequestPayload
  [SchaltEvent.CloneProgress]: CloneProgressPayload
  [SchaltEvent.OrchestratorLaunchFailed]: OrchestratorLaunchFailedPayload
  [SchaltEvent.DiffBaseBranchChanged]: DiffBaseBranchChangedPayload
  [SchaltEvent.ProjectValidationError]: ProjectValidationErrorPayload
  [SchaltEvent.OpenPrModal]: OpenPrModalPayload
  [SchaltEvent.OpenMergeModal]: OpenMergeModalPayload
  [SchaltEvent.OpenGitlabMrModal]: OpenGitlabMrModalPayload
  [SchaltEvent.SelectAllRequested]: null
}

File: src-tauri/src/events.rs
use serde::{Deserialize, Serialize};
use tauri::Emitter;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum SchaltEvent {
    SessionsRefreshed,
    SessionAdded,
    SessionRemoved,
    SessionCancelling,
    CancelError,
    SessionActivity,
    SessionGitStats,
    TerminalAttention,
    TerminalClosed,
    TerminalForceScroll,
    TerminalOutputChanged,
    PtyData,
    GlobalKeepAwakeStateChanged,
    ProjectReady,
    OpenDirectory,
    OpenHome,
    FileChanges,
    FollowUpMessage,
    Selection,
    GitOperationStarted,
    GitOperationCompleted,
    GitOperationFailed,
    ProjectFilesUpdated,
    GitHubStatusChanged,
    AppUpdateResult,
    DevBackendError,
    SetupScriptRequested,
    CloneProgress,
    OrchestratorLaunchFailed,
    ProjectValidationError,
    OpenPrModal,
    OpenMergeModal,
    OpenGitlabMrModal,
    SelectAllRequested,
}

impl SchaltEvent {
    pub fn as_str(&self) -> &'static str {
        match self {
            SchaltEvent::SessionsRefreshed => "schaltwerk:sessions-refreshed",
            SchaltEvent::SessionAdded => "schaltwerk:session-added",
            SchaltEvent::SessionRemoved => "schaltwerk:session-removed",
            SchaltEvent::SessionCancelling => "schaltwerk:session-cancelling",
            SchaltEvent::CancelError => "schaltwerk:cancel-error",
            SchaltEvent::SessionActivity => "schaltwerk:session-activity",
            SchaltEvent::SessionGitStats => "schaltwerk:session-git-stats",
            SchaltEvent::TerminalAttention => "schaltwerk:terminal-attention",
            SchaltEvent::TerminalClosed => "schaltwerk:terminal-closed",
            SchaltEvent::TerminalForceScroll => "schaltwerk:terminal-force-scroll",
            SchaltEvent::TerminalOutputChanged => "schaltwerk:terminal-output-changed",
            SchaltEvent::PtyData => "schaltwerk:pty-data",
            SchaltEvent::GlobalKeepAwakeStateChanged => {
                "schaltwerk:global-keep-awake-state-changed"
            }
            SchaltEvent::ProjectReady => "schaltwerk:project-ready",
            SchaltEvent::OpenDirectory => "schaltwerk:open-directory",
            SchaltEvent::OpenHome => "schaltwerk:open-home",
            SchaltEvent::FileChanges => "schaltwerk:file-changes",
            SchaltEvent::FollowUpMessage => "schaltwerk:follow-up-message",
            SchaltEvent::Selection => "schaltwerk:selection",
            SchaltEvent::GitOperationStarted => "schaltwerk:git-operation-started",
            SchaltEvent::GitOperationCompleted => "schaltwerk:git-operation-completed",
            SchaltEvent::GitOperationFailed => "schaltwerk:git-operation-failed",
            SchaltEvent::ProjectFilesUpdated => "schaltwerk:project-files-updated",
            SchaltEvent::GitHubStatusChanged => "schaltwerk:github-status-changed",
            SchaltEvent::AppUpdateResult => "schaltwerk:app-update-result",
            SchaltEvent::DevBackendError => "schaltwerk:dev-backend-error",
            SchaltEvent::SetupScriptRequested => "schaltwerk:setup-script-request",
            SchaltEvent::CloneProgress => "schaltwerk:clone-progress",
            SchaltEvent::OrchestratorLaunchFailed => "schaltwerk:orchestrator-launch-failed",
            SchaltEvent::ProjectValidationError => "schaltwerk:project-validation-error",
            SchaltEvent::OpenPrModal => "schaltwerk:open-pr-modal",
            SchaltEvent::OpenMergeModal => "schaltwerk:open-merge-modal",
            SchaltEvent::OpenGitlabMrModal => "schaltwerk:open-gitlab-mr-modal",
            SchaltEvent::SelectAllRequested => "schaltwerk:select-all-requested",
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum CloneProgressKind {
    Info,
    Success,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CloneProgressPayload {
    pub request_id: String,
    pub message: String,
    pub remote: String,
    pub kind: CloneProgressKind,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectValidationErrorPayload {
    pub path: String,
    pub error: String,
}

pub fn emit_event<T: Serialize + Clone>(
    app: &tauri::AppHandle,
    event: SchaltEvent,
    payload: &T,
) -> Result<(), tauri::Error> {
    app.emit(event.as_str(), payload)
}
